File: python/src/fise/_lz4.py
from __future__ import annotations

from .errors import FiseError

_MIN_MATCH = 4
_LAST_LITERALS = 5
_MATCH_FIND_LIMIT = 12
_MAX_DISTANCE = 0xFFFF
_HASH_SIZE = 1 << 16
_HASH_MULTIPLIER = 0x9E3779B1


def compress_lz4_block(data: bytes) -> bytes:
    positions = [-1] * _HASH_SIZE
    output = bytearray()
    cursor = 0
    anchor = 0
    match_find_limit = len(data) - _MATCH_FIND_LIMIT
    match_limit = len(data) - _LAST_LITERALS

    while cursor <= match_find_limit:
        sequence = _read_uint32(data, cursor)
        hash_value = ((sequence * _HASH_MULTIPLIER) & 0xFFFFFFFF) >> 16
        candidate = positions[hash_value]
        positions[hash_value] = cursor
        if (
            candidate < 0
            or cursor - candidate > _MAX_DISTANCE
            or _read_uint32(data, candidate) != sequence
        ):
            cursor += 1
            continue

        match_start = cursor
        candidate_cursor = candidate + _MIN_MATCH
        cursor += _MIN_MATCH
        while cursor < match_limit and data[cursor] == data[candidate_cursor]:
            cursor += 1
            candidate_cursor += 1
        literal_length = match_start - anchor
        match_length = cursor - match_start
        token_index = len(output)
        output.append(0)
        output[token_index] = (min(literal_length, 15) << 4) | min(match_length - _MIN_MATCH, 15)
        if literal_length >= 15:
            _write_extended_length(output, literal_length - 15)
        output.extend(data[anchor:match_start])
        distance = match_start - candidate
        output.append(distance & 0xFF)
        output.append(distance >> 8)
        encoded_match_length = match_length - _MIN_MATCH
        if encoded_match_length >= 15:
            _write_extended_length(output, encoded_match_length - 15)
        anchor = cursor

    literal_length = len(data) - anchor
    output.append(min(literal_length, 15) << 4)
    if literal_length >= 15:
        _write_extended_length(output, literal_length - 15)
    output.extend(data[anchor:])
    return bytes(output)


def decompress_lz4_block(data: bytes, expected_length: int) -> bytes:
    if type(expected_length) is not int or expected_length < 0:
        raise _invalid_block("declared output length is invalid")
    try:
        output = bytearray(expected_length)
    except (MemoryError, OverflowError) as error:
        raise _invalid_block("declared output length cannot be allocated", error) from error
    input_cursor = 0
    output_cursor = 0
    final_literal_length = -1
    last_match_start = -1

    while input_cursor < len(data):
        token = data[input_cursor]
        input_cursor += 1
        literal_length = token >> 4
        if literal_length == 15:
            literal_length, input_cursor = _read_extended_length(
                data,
                input_cursor,
                15,
                expected_length - output_cursor,
            )
        if (
            literal_length > expected_length - output_cursor
            or literal_length > len(data) - input_cursor
        ):
            raise _invalid_block("literal run exceeds its input or output boundary")
        output[output_cursor:output_cursor + literal_length] = data[
            input_cursor:input_cursor + literal_length
        ]
        input_cursor += literal_length
        output_cursor += literal_length

        if input_cursor == len(data):
            final_literal_length = literal_length
            break
        if len(data) - input_cursor < 2:
            raise _invalid_block("match offset is truncated")
        distance = data[input_cursor] | (data[input_cursor + 1] << 8)
        input_cursor += 2
        if distance == 0 or distance > output_cursor:
            raise _invalid_block("match offset points outside restored data")

        encoded_match_length = token & 0x0F
        if encoded_match_length == 15:
            encoded_match_length, input_cursor = _read_extended_length(
                data,
                input_cursor,
                15,
                expected_length - output_cursor - _MIN_MATCH,
            )
        match_length = encoded_match_length + _MIN_MATCH
        if match_length > expected_length - output_cursor:
            raise _invalid_block("match run exceeds the declared output length")
        last_match_start = output_cursor
        source_cursor = output_cursor - distance
        for index in range(match_length):
            output[output_cursor + index] = output[source_cursor + index]
        output_cursor += match_length

    if (
        input_cursor != len(data)
        or output_cursor != expected_length
        or final_literal_length < 0
    ):
        raise _invalid_block("block does not restore the declared output length")
    if last_match_start >= 0 and (
        final_literal_length < _LAST_LITERALS
        or last_match_start > expected_length - _MATCH_FIND_LIMIT
    ):
        raise _invalid_block("block violates the LZ4 terminal sequence boundary")
    return bytes(output)


def _read_uint32(data: bytes, offset: int) -> int:
    return (
        data[offset]
        | (data[offset + 1] << 8)
        | (data[offset + 2] << 16)
        | (data[offset + 3] << 24)
    )


def _write_extended_length(output: bytearray, length: int) -> None:
    while length >= 255:
        output.append(255)
        length -= 255
    output.append(length)


def _read_extended_length(
    data: bytes,
    offset: int,
    base: int,
    maximum: int,
) -> tuple[int, int]:
    length = base
    while True:
        if offset >= len(data):
            raise _invalid_block("extended length is truncated")
        value = data[offset]
        offset += 1
        if maximum < 0 or length > maximum - value:
            raise _invalid_block("extended length exceeds its boundary")
        length += value
        if value != 255:
            return length, offset


def _invalid_block(reason: str, cause: BaseException | None = None) -> FiseError:
    return FiseError(
        "INVALID_PAYLOAD",
        f"FISE: invalid compressed structured payload; {reason}.",
        cause,
    )

File: src/v2/lz4.ts
import { FiseError } from "../errors.js";

const MIN_MATCH = 4;
const LAST_LITERALS = 5;
const MATCH_FIND_LIMIT = 12;
const MAX_DISTANCE = 0xffff;
const HASH_BITS = 16;
const HASH_SIZE = 1 << HASH_BITS;
const HASH_SHIFT = 32 - HASH_BITS;
const HASH_MULTIPLIER = 0x9e37_79b1;

/** Encodes one deterministic independent LZ4 block. */
export function compressLz4Block(input: Uint8Array): Uint8Array {
	const output = new Uint8Array(compressBound(input.length));
	const positions = new Int32Array(HASH_SIZE);
	positions.fill(-1);
	let inputOffset = 0;
	let anchor = 0;
	let outputOffset = 0;
	const matchFindLimit = input.length - MATCH_FIND_LIMIT;
	const matchLimit = input.length - LAST_LITERALS;

	while (inputOffset <= matchFindLimit) {
		const sequence = readUint32(input, inputOffset);
		const hash = Math.imul(sequence, HASH_MULTIPLIER) >>> HASH_SHIFT;
		const candidate = positions[hash];
		positions[hash] = inputOffset;
		if (
			candidate < 0 ||
			inputOffset - candidate > MAX_DISTANCE ||
			readUint32(input, candidate) !== sequence
		) {
			inputOffset++;
			continue;
		}

		const matchStart = inputOffset;
		let candidateOffset = candidate + MIN_MATCH;
		inputOffset += MIN_MATCH;
		while (
			inputOffset < matchLimit &&
			input[inputOffset] === input[candidateOffset]
		) {
			inputOffset++;
			candidateOffset++;
		}
		const literalLength = matchStart - anchor;
		const matchLength = inputOffset - matchStart;
		const tokenOffset = outputOffset++;
		output[tokenOffset] =
			(Math.min(literalLength, 15) << 4) |
			Math.min(matchLength - MIN_MATCH, 15);
		if (literalLength >= 15) {
			outputOffset = writeExtendedLength(output, outputOffset, literalLength - 15);
		}
		output.set(input.subarray(anchor, matchStart), outputOffset);
		outputOffset += literalLength;
		const distance = matchStart - candidate;
		output[outputOffset++] = distance & 0xff;
		output[outputOffset++] = distance >>> 8;
		const encodedMatchLength = matchLength - MIN_MATCH;
		if (encodedMatchLength >= 15) {
			outputOffset = writeExtendedLength(
				output,
				outputOffset,
				encodedMatchLength - 15
			);
		}
		anchor = inputOffset;
	}

	const literalLength = input.length - anchor;
	const tokenOffset = outputOffset++;
	output[tokenOffset] = Math.min(literalLength, 15) << 4;
	if (literalLength >= 15) {
		outputOffset = writeExtendedLength(output, outputOffset, literalLength - 15);
	}
	output.set(input.subarray(anchor), outputOffset);
	outputOffset += literalLength;
	return output.slice(0, outputOffset);
}

/** Restores one independent LZ4 block into an exact bounded output. */
export function decompressLz4Block(
	input: Uint8Array,
	expectedLength: number
): Uint8Array {
	if (!Number.isSafeInteger(expectedLength) || expectedLength < 0) {
		throw invalidBlock("declared output length is invalid");
	}
	const output = new Uint8Array(expectedLength);
	let inputOffset = 0;
	let outputOffset = 0;
	let finalLiteralLength = -1;
	let lastMatchStart = -1;

	while (inputOffset < input.length) {
		const token = input[inputOffset++];
		let literalLength = token >>> 4;
		if (literalLength === 15) {
			[literalLength, inputOffset] = readExtendedLength(
				input,
				inputOffset,
				15,
				expectedLength - outputOffset
			);
		}
		if (
			literalLength > expectedLength - outputOffset ||
			literalLength > input.length - inputOffset
		) {
			throw invalidBlock("literal run exceeds its input or output boundary");
		}
		output.set(input.subarray(inputOffset, inputOffset + literalLength), outputOffset);
		inputOffset += literalLength;
		outputOffset += literalLength;

		if (inputOffset === input.length) {
			finalLiteralLength = literalLength;
			break;
		}
		if (input.length - inputOffset < 2) {
			throw invalidBlock("match offset is truncated");
		}
		const distance = input[inputOffset] | (input[inputOffset + 1] << 8);
		inputOffset += 2;
		if (distance === 0 || distance > outputOffset) {
			throw invalidBlock("match offset points outside restored data");
		}

		let encodedMatchLength = token & 0x0f;
		if (encodedMatchLength === 15) {
			[encodedMatchLength, inputOffset] = readExtendedLength(
				input,
				inputOffset,
				15,
				expectedLength - outputOffset - MIN_MATCH
			);
		}
		const matchLength = encodedMatchLength + MIN_MATCH;
		if (matchLength > expectedLength - outputOffset) {
			throw invalidBlock("match run exceeds the declared output length");
		}
		lastMatchStart = outputOffset;
		const sourceOffset = outputOffset - distance;
		for (let index = 0; index < matchLength; index++) {
			output[outputOffset + index] = output[sourceOffset + index];
		}
		outputOffset += matchLength;
	}

	if (inputOffset !== input.length || outputOffset !== expectedLength || finalLiteralLength < 0) {
		throw invalidBlock("block does not restore the declared output length");
	}
	if (
		lastMatchStart >= 0 &&
		(finalLiteralLength < LAST_LITERALS || lastMatchStart > expectedLength - MATCH_FIND_LIMIT)
	) {
		throw invalidBlock("block violates the LZ4 terminal sequence boundary");
	}
	return output;
}

function readUint32(input: Uint8Array, offset: number): number {
	return (
		input[offset] |
		(input[offset + 1] << 8) |
		(input[offset + 2] << 16) |
		(input[offset + 3] << 24)
	) >>> 0;
}

function writeExtendedLength(
	output: Uint8Array,
	offset: number,
	length: number
): number {
	while (length >= 255) {
		output[offset++] = 255;
		length -= 255;
	}
	output[offset++] = length;
	return offset;
}

function readExtendedLength(
	input: Uint8Array,
	offset: number,
	base: number,
	maximum: number
): [number, number] {
	let length = base;
	while (true) {
		if (offset >= input.length) throw invalidBlock("extended length is truncated");
		const value = input[offset++];
		if (length > maximum - value) throw invalidBlock("extended length exceeds its boundary");
		length += value;
		if (value !== 255) return [length, offset];
	}
}

function compressBound(length: number): number {
	if (!Number.isSafeInteger(length) || length < 0) {
		throw new FiseError("INVALID_INPUT", "FISE: structured payload length is invalid.");
	}
	const bound = length + Math.floor(length / 255) + 16;
	if (!Number.isSafeInteger(bound)) {
		throw new FiseError("ENVELOPE_LIMIT", "FISE: structured payload is too large.");
	}
	return bound;
}

function invalidBlock(reason: string): FiseError {
	return new FiseError("INVALID_PAYLOAD", `FISE: invalid compressed structured payload; ${reason}.`);
}
